lib/module.nix

```diff
@@ -0,0 +1,376 @@
+
+  # Global database lookup shortcuts:
+  dbSite = db.sites."${site}";
+
+  # Supervisor build infrastructure:
+  fenix = import (builtins.fetchGit {
+    url = "https://github.com/nix-community/fenix.git";
+    ref = "main";
+    rev = "0900ff903f376cc822ca637fef58c1ca4f44fab5";
+  }) { };
+
+  tmlSource = builtins.fetchGit {
+    url = "https://github.com/treadmill-tb/treadmill";
+    ref = "dev/qemu-tracing";
+    rev = "ddbba1723977b320b6e0ec72342cf0527b2a9736";
+  };
+
+  rustToolchain = fenix.fromToolchainFile {
+    file = "${tmlSource}/rust-toolchain.toml";
+    sha256 = "sha256-6eN/GKzjVSjEhGO9FhWObkRFaE1Jf+uqMSdQnb8lcB4=";
+  };
+
+  rustPlatform = pkgs.makeRustPlatform {
+    cargo = rustToolchain;
+    rustc = rustToolchain;
+  };
+
+  tmlQemuSupervisor = rustPlatform.buildRustPackage rec {
+    pname = "tml-qemu-supervisor";
+    version = "0.0.1";
+
+    src = "${tmlSource}";
+    buildAndTestSubdir = "supervisor/qemu";
+
+    cargoLock.lockFile = "${src}/Cargo.lock";
+    cargoLock.outputHashes."inquire-0.7.5" = "sha256-iEdsjq4IYYl6QoJmDkPQS5bJJvPG3sehDygefAOhTrY=";
+  };
+
+  # Configuration generators:
+  genNbdNetbootSupervisorConfig = supervisorId: let
+    dbSupervisor = db.supervisors."${supervisorId}";
+    dbSiteSupervisor = dbSite.supervisors."${supervisorId}";
+    dbSwitch = dbSite.switches."${dbSiteSupervisor.nbd_netboot_host_switch}";
+
+    vlanNetdevName = "tn-${shortId supervisorId}";
+    networkCIDR = "${dbSiteSupervisor.nbd_netboot_host_ip4.network}/${toString dbSiteSupervisor.nbd_netboot_host_ip4.prefixlen}";
+    supervisorCIDR = "${dbSiteSupervisor.nbd_netboot_host_ip4.supervisor_addr}/${toString dbSiteSupervisor.nbd_netboot_host_ip4.prefixlen}";
+    hostCIDR = "${dbSiteSupervisor.nbd_netboot_host_ip4.addr}/${toString dbSiteSupervisor.nbd_netboot_host_ip4.prefixlen}";
+  in {
+    # VLAN network device on the experiment link
+    systemd.network.netdevs."10-${vlanNetdevName}" = {
+      netdevConfig = {
+        Name = "${vlanNetdevName}";
+        Kind = "vlan";
+      };
+      vlanConfig.Id = dbSiteSupervisor.nbd_netboot_host_vlan;
+    };
+
+    # Add this VLAN to the trunk link.
+    #
+    # TODO: this is not nice. We expect that the importing system configurations
+    # defines a systemd-networkd network definition named the same as the
+    # underlying network interface. Is there a better way to merge
+    # configurations like this?
+    systemd.network.networks."${dbSwitch.trunk_supervisor_netdev}" = {
+      networkConfig.VLAN = [ "${vlanNetdevName}" ];
+    };
+
+    # VLAN network configuration:
+    systemd.network.networks."20-${vlanNetdevName}" =
+      if dbSupervisor.nbd_netboot_host.pxe_profile == "raspberrypi" then {
+        matchConfig.Name = vlanNetdevName;
+        address = [ supervisorCIDR ];
+        linkConfig.RequiredForOnline = "no";
+        networkConfig.DHCPServer = "yes";
+        dhcpServerConfig = {
+          EmitDNS = "yes";
+          #DNS = [ "10.64.0.1" ];
```

Comment:
Remove. We may want to make the DNS server configurable, as it depends on the upstream configuration. Using Google DNS seems fine for now, though.